enhancements/ingress/configurable-dns-loadbalancerservice.md

```diff
@@ -0,0 +1,316 @@
+  // operator and should be managed at the discretion of the cluster admin.
+  //
+  // +kubebuilder:default:="Managed"
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:Enum=Managed;Unmanaged
+  // +optional
+  ManagementPolicy DNSManagementPolicy `json:"managementPolicy"`
+}
+
+// DNSManagementPolicy is a policy for configuring how the dns controller 
+// manages DNSRecords.
+//
+// +kubebuilder:validation:Enum=Managed;Unmanaged
+type DNSManagementPolicy string
+
+const (
+    // ManagedDNS configures the dns controller manage the lifecyle of the
+    // dns record on the appropriate platform.
+    ManagedDNS DNSManagementPolicy = "Managed"
+    // UnmanagedDNS configures the dns controller to ignore the dns record
+    // and allows to be manually managed by the cluster admin.
+    UnmanagedDNS DNSManagementPolicy = "Unmanaged"
+)
+```
+
+Re-uses existing condition `DNSManaged` of the controller to denote if
+the wildcard *DNSRecord* associated with the controller is managed/unmanaged. 
+
+### Implementation Details/Notes/Constraints
+
+Based on `.loadBalancer.dns`, the controller decides whether to ensure creating
+a wildcard *DNSRecord*. If the customer is updating `.loadBalancer.dns` from 
+`"Managed"` to `"Unmanaged"` the previously created *DNSRecord* will be updated
+appropriately.
+
+The existing state of the ingresscontroller will be synced to the *DNSRecord* as
+well, to clearly state which dns record is unmanaged. The cluster admin can choose
+to delete the *DNSRecord* at his own discretion. 
+
+Once the *DNSRecord* is set to `Unmanaged`, any updates done to it will be ignored
+by the operator, but it to be noted that if the ingresscontroller is updated from
+`Unmanaged` to `Managed` any changes done will be lost during re-sync.
+
+The support provided by the installer and this new field aren't entirely
+mutually exclusive, i.e. if customer has opted to disable DNS management cluster
+wide (as documented
+[here](https://github.com/openshift/installer/blob/master/docs/user/aws/install_upi.md#remove-dns-zones)),
+this field is of no value and need not be set since *DNSRecord* created is silently
+reconciled by the DNS controller.
+
+### Risks and Mitigations
+
+This feature does not support default ingresscontrollers, if updated to be unmanaged
+in conjunction with updating the domain could result in breaking cluster connectivity
+if not all components are properly updated. This should be done at the discretion
+of the cluster admin. The manual creation of the required DNS records will need to be
+done prior to making this change, as the canary controller will try to ensure that the
+ingress domains are reachable, and otherwise this sets the `cluster-ingress-operator`
+in a degraded state.
+
+### Drawbacks
+
+After an ingresscontroller is updated from `Managed` to `Unmanaged`, the previous
+*DNSRecord* is left in an `Unmanaged` state which could result in orphaned resources
```

Comment:
I'm curious about what "orphaned" means in this context. Does it mean that if your IC's `.spec.dns` is set to `Unmanaged` and you delete your IC that the `DNSRecord` CRs will stay on the cluster? If that's the case, this may be a good opportunity for an alert indicating that there are orphaned resources in cases where you have DNSRecord objects with no corresponding IC. I haven't looked too closely at whether that can happen though so it'd be worth validating first.

Reply:
Yes, your understanding is correct. So by alert do you mean some kind of `Event` or do you mean a `Log`? (Both are feasible)

Reply:
Looking at the existing alerts, they seem to be very critical in nature, i.e. scenarios where the ingresscontroller is completely non-functional or where it has incorrect configs. In contrast to that, having an alert for orphaned *DNSRecord*s can't really be justified when these orphaned resources don't impact the functioning of the ingresscontroller. 

One of the main reasons we decided not to delete the resource was to enable the cluster admin to create the required DNS record manually looking at the *DNSRecord* (has all the required information, otherwise would need to gather information from the routes and ingresscontroller).

Reply:
> having an alert for orphaned DNSRecords can't really be justified when these orphaned resources don't impact the functioning of the ingresscontroller.

I think we're coming from different perspectives and that's maybe muddying the water a bit. From an operator perspective, I agree with you. No need to alert on the operator if the operator is functioning normally. However, consider the customer or SD perspective when an IC is accidentally deleted or they discover (without an alert) that they unexpectedly have _n_ orphaned DNSRecords. Maybe the middle-ground in this is to expose the metric so it can be captured and queried but do not create an alert. Then we're solving both needs by making the orphaned DNSRecord count easily retrievable without necessarily indicating to cluster admins that they need to take urgent action. What do you think?

I'd be curious whether @Miciah has an opinion on this.

Reply:
I'm still not very sure as to what problem we are attempting to solve with this metric. Are we trying to solve the problem where cluster admins forget to clean up the *DNSRecord* post migrating from `Managed` to `Unmanaged` or are we looking for a way to provide more information to cluster admins about the current state of ingresscontroller (especially about orphaned *DNSRecord* count)?  

If it is the latter, and we want to give more information about the *DNSRecords* (counts) it still doesn't seem to justify a metric, especially when the transition from `Managed` to `Unmanaged` is more of a one-time operation (I don't see customers frequently toggling this). And if we are talking about exposing other details (`IC is accidentally deleted`) better to take that up separately. 

If it is the former, I think documentation is sufficient, since this procedure of migrating from `Managed` to `Unmanaged` is very infrequent. We can maybe have `WARN` logs on the controller where we ignore the *DNSRecord* to nudge the cluster admin to clean up.

Reply:
Stepping back a bit, if the question from a cluster admin is "How do I see if I have orphaned DNSRecord objects?" How would you answer that question?

Reply:
The cluster admin can fetch all *DNSRecords* and filter where `managementPolicy` is set to `Unmanaged`. 
```bash
oc get dnsrecord -A  -o=jsonpath="{.items[?(@.spec.managementPolicy=='Unmanaged')].metadata.name}"
``` 
Maybe we can introduce a new status into *DNSRecord*, which can be used as filter.

Reply:
@thejasn thanks for the time on the call! It sounds like the current implementation allows for DNSRecords to be orphaned when an IC is deleted (correct me if I'm wrong) so in order to avoid that, the implementation of this EP will ensure that DNSRecords do not get orphaned when an IC is deleted. That'll remove the need for a metric/alert around orphaned DNSRecord objects and result in a better UX since we won't be leaving objects around. One question this raises for me though is, do we have any use case for leaving the DNSRecord around in an orphaned state?

Reply:
@brandisher Existing implementation doesn't allow orphaned DNSRecords, the ingresscontroller adds a finalizer to the *DNSRecord*, so that it can consequently be cleaned up during deletion of the IC. This EP will introduce the behavior where we would end up with *DNSRecords* which aren't associated with any IC (orphaned), these can be queried by the cluster admin as shown above  and proceed to delete. 

We will ensure the cluster admin deletes the *DNSRecord* by not removing the finalizer on the *DNSRecord* when the IC was marked for deletion. So the workflow would be something like, 
1. Set the field to `managementPolicy: Unmanaged`
2. Later on, when the admin deletes IC (tries to orphan the *DNSRecord*)
3. Delete the finalizer on the *DNSRecord* manually so that IC can also be deleted. 
This would ensure the there are no orphaned resources. 

Regarding the adding a status/condition on *DNSRecord*, had a discussion with @Miciah, and we concluded that querying using the spec is sufficient and adding a new condition is redundant since the content is pretty much the same.